backend/service/auditsink/slack/slack_test.go

```diff
@@ -51,3 +55,187 @@ func TestFormat(t *testing.T) {
+	anyEc2Req, _ := anypb.New(&ec2v1.ResizeAutoscalingGroupRequest{Size: &ec2v1.AutoscalingGroupSize{Min: 2, Max: 4, Desired: 3}})
+	anyEc2Resp, _ := anypb.New(&ec2v1.ResizeAutoscalingGroupResponse{})
+
+	anyK8sDescribeReq, _ := anypb.New(&k8sapiv1.DescribePodRequest{Name: "foo"})
+	anyK8sDescribeResp, _ := anypb.New(&k8sapiv1.DescribePodResponse{Pod: &k8sapiv1.Pod{PodIp: "000"}})
+
+	k8sUpdateReq := &k8sapiv1.UpdatePodRequest{
+		ExpectedObjectMetaFields: &k8sapiv1.ExpectedObjectMetaFields{
+			Annotations: map[string]*k8sapiv1.NullableString{
+				"foo": &k8sapiv1.NullableString{Kind: &k8sapiv1.NullableString_Value{Value: "new-value"}},
+			},
+		},
+		ObjectMetaFields:       &k8sapiv1.ObjectMetaFields{Labels: map[string]string{"foo": "new-value"}},
+		RemoveObjectMetaFields: &k8sapiv1.RemoveObjectMetaFields{Annotations: []string{"foo", "bar"}},
+	}
+
+	anyK8sUpdateReq, _ := anypb.New(k8sUpdateReq)
+	anyK8UpdateResp, _ := anypb.New(&k8sapiv1.UpdatePodResponse{})
+
+	testCases := []struct {
+		text           string
+		event          *auditv1.RequestEvent
+		expectedErr    bool
+		expectedOutput string
+	}{
+		// metdata from the API request
+		{
+			text: "`Min size` is {{.Request.size.min}}, `Max size` is {{.Request.size.max}}, `Desired size` is {{.Request.size.desired}}",
+			event: &auditv1.RequestEvent{
+				RequestMetadata:  &auditv1.RequestMetadata{Body: anyEc2Req},
+				ResponseMetadata: &auditv1.ResponseMetadata{Body: anyEc2Resp},
+			},
+			expectedOutput: "`Min size` is 2, `Max size` is 4, `Desired size` is 3",
+		},
+		// metadata from both the API request and repsonse
+		{
+			text: "{{.Request.name}} ip address is {{.Response.pod.podIp}}",
+			event: &auditv1.RequestEvent{
+				RequestMetadata:  &auditv1.RequestMetadata{Body: anyK8sDescribeReq},
+				ResponseMetadata: &auditv1.ResponseMetadata{Body: anyK8sDescribeResp},
+			},
+			expectedOutput: "foo ip address is 000",
+		},
+		// metadata that is a map, uses helper slackList
+		{
+			text: "*Updated labels*:{{slackList .Request.objectMetaFields.labels}}",
+			event: &auditv1.RequestEvent{
+				RequestMetadata:  &auditv1.RequestMetadata{Body: anyK8sUpdateReq},
+				ResponseMetadata: &auditv1.ResponseMetadata{Body: anyK8UpdateResp},
+			},
+			expectedOutput: "*Updated labels*:\n- foo: new-value",
+		},
+		// metdata that is a list, uses helper slackList
+		{
+			text: "*Removed annotations*:{{slackList .Request.removeObjectMetaFields.annotations}}",
+			event: &auditv1.RequestEvent{
+				RequestMetadata:  &auditv1.RequestMetadata{Body: anyK8sUpdateReq},
+				ResponseMetadata: &auditv1.ResponseMetadata{Body: anyK8UpdateResp},
+			},
+			expectedOutput: "*Removed annotations*:\n- foo\n- bar",
+		},
+		// metdata that is a map, map value is a another map
+		// uses the Golang template `range`
+		{
+			text: "*Expected Preconditions*:{{range $key, $val := .Request.expectedObjectMetaFields.annotations}}\n- {{$key}}: {{range $i, $j := $val}}{{$j}}{{end}}{{end}}",
```

Comment:
this is the `range` syntax for https://github.com/lyft/clutch/blob/main/api/k8s/v1/k8s.proto#L333 which is a `map<string, message> type`